src/Lean/Elab/Coinductive.lean

```diff
@@ -0,0 +1,467 @@
+      -/
+      let goalType := mkConst (removeFunctorPostfix name) levelParams
+      let mut goalType := mkAppN goalType params
+      goalType := mkAppN goalType bodyAppArgs
+      trace[Elab.coinductive] "The conclusion of the constructor {ctor.name} is {goalType}"
+
+      -- We start by making the metavariable for it, that we will fill
+      let goal ← mkFreshExprMVar <| .some goalType
+      let hole := Expr.mvarId! goal
+      let unfoldEq := mkConst ((removeFunctorPostfix name) ++ `functor_unfold) levelParams
+      let unfoldEq := mkAppN unfoldEq params
+
+      let rewriteResult ← hole.rewrite (←hole.getType) unfoldEq
+
+      let newHole ← hole.replaceTargetEq rewriteResult.eNew rewriteResult.eqProof
+
+      /-
+        Now, all it suffices is to call an approprate constructor of the flat inductive.
+      -/
+      let constructor := mkConst ctor.name levelParams
+      let constructor := mkAppN constructor params
+      let constructor := mkAppN constructor predicates
+      let constructor := mkAppN constructor bodyArgs
+      newHole.assign constructor
+      let conclusion ← instantiateMVars goal
+      let conclusion ← mkLambdaFVars bodyArgs conclusion
+      mkLambdaFVars params conclusion
+    let type ← inferType res
+    trace[Elab.coinductive] "The elaborated constructor is of the type: {type}"
+    /-
+      We finish by registering the appropriate declaration
+    -/
+    addDecl <| .defnDecl (←mkDefinitionValInferringUnsafe
+      (removeFunctorPostfixInCtor ctor.name) ctor.levelParams type res .opaque)
+    Term.addTermInfo' ctorSyntax res (isBinder := true)
+
+/--
+  Given the number of parameters and the `InductiveVal` containing flat inductives
+  (see `mkFlatInductive`) and `CoinductiveElabData` associated with the mutual coinductive
+  predicates, generates their constructors that correspond to the
+  constructors given in the original syntax.
+-/
+private def generateCoinductiveConstructors (numParams : Nat) (infos : Array InductiveVal)
+    (coinductiveElabData : Array CoinductiveElabData) : TermElabM Unit := do
+  for indType in infos, e in coinductiveElabData do
+    for ctor in indType.ctors, ctorSyntax in e.ctorSyntax do
+      generateCoinductiveConstructor infos ctorSyntax numParams indType.name
+        <| ←getConstInfoCtor ctor
+
+/--
+  Generates `casesOn` eliminators for coinductive predicates.
+  These eliminators allow pattern matching on coinductive predicates,
+  enabling case analysis in proofs.
+-/
+private def mkCasesOnCoinductive (infos : Array InductiveVal) : MetaM Unit := do
+  let levels := infos[0]!.levelParams.map mkLevelParam
+  let allCtors := infos.flatMap (·.ctors.toArray)
+
+  forallBoundedTelescope infos[0]!.type (infos[0]!.numParams - infos.size) fun params _ => do
+  let predicates := infos.map fun info => mkConst (removeFunctorPostfix info.name) levels
+  let predicates := predicates.map (mkAppN · params)
+  for info in infos do
+    let casesOnName := (info.name ++ `casesOn)
```

Comment:
There is `Lean.mkCasesOnName`. And it doesn’t use `++` but `Name.str`, which makes me wonder if one of the two variants is broken in the presence of macro scopes. (Because `++` on `Name` appends underneath the macroscope, see [docstrings](https://leanprover-community.github.io/mathlib4_docs/Init/Prelude.html#Lean.Name.append), but `Name.str` appends at the end. Oh wey).